File: LeetCode/Python/swap_nodes_in_pairs.py
#!/usr/bin/env python
#coding: utf-8

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        k = ListNode(-1)
        k.next = head
        tk = k
        while tk.next:
            n1 = tk.next
            n2 = n1.next
            if n2:
                n1.next = n2.next
                n2.next = n1
                tk.next = n2
            tk = n1
        return k.next



def getList(m, n):
    h = th = ListNode(m)
    for i in range(m+1, n+1):
        th.next = ListNode(i)
        th = th.next
    return h

def printList(head):
    while head:
        print(head.val),
        head = head.next
    print('\n')

if __name__ == '__main__':
    s = Solution()
    head = getList(1,1)
    printList(head)
    printList(s.swapPairs(head))

File: LeetCode/Java/Solution/src/ios/ac/cn/SwapNodesInPairs.java
package ios.ac.cn;

public class SwapNodesInPairs {
	private ListNode modify(ListNode start) {
		if (start == null || start.next == null) {
			return start;
		}

		ListNode startNext = start.next;
		start.next = startNext.next;
		startNext.next = start;
		return startNext;
	}

	public ListNode swapPairs(ListNode head) {
		head = modify(head);

		ListNode cur = head;
		while (cur != null && cur.next != null && cur.next.next != null) {
			ListNode next = modify(cur.next.next);
			cur.next.next = next;
			cur = next;
		}

		return head;
	}
}
